### PR TITLE
Adding FieldValue.numericAdd()

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR adds support for the Numeric Add transform and follows the semantics of the backend implementation. To avoid flicker with latency compensation, it adds the concept of "base values", which are stored as a list of patch mutations along with the transform mutations in a mutation batch. The base mutations are applied first and are used to overwrite values from the remote document cache.

~~This PR also simplifies some plumbing in regards to the "previous value", which is  used when rendering Server Timestamps. Before, if two server timestamps replaced the same value in sequence, each timestamp would point to this previous value. Now, the previous value is chained, and the second timestamp obtains its previous value by asking the first timestamp for its previous value. This is slightly less optimal, but simplifies the changes that this PR introduces to Local Store.~~

As an additional benefit, and to get this PR passed 1000 lines of code, I added a flag to run the Integration Tests against Hexa. Because I don't know how to add custom configurations in gradle, this is for now hardcoded.

### CODE REVIEW

**Comment by schmidt-sebastian** [comment]
Hm, looks like this broke `testServerTimestampsCanReturnPreviousValue`. I'll take a look.

**Comment by schmidt-sebastian** [comment]
> Hm, looks like this broke `testServerTimestampsCanReturnPreviousValue`. I'll take a look.

I reverted the change that simplified the plumbing as this broke updates for transforms that were nested inside of a map.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java [hunk 1]
```diff
@@ -121,6 +121,25 @@ public MaybeDocument applyToLocalView(
...
+
+  @Override
+  public boolean isIdempotent() {
+    for (FieldTransform transform : fieldTransforms) {
+      if (!transform.isIdempotent()) {
+        return false;
+      }
+    }
+    return true;
+  }
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java]: **Update:** I see that I'm wrong (I forgot there's a separate TransformMutation for every field with a transform) and that you have a test for this. Nice work!

<s>It occurs to me that we treat the entire mutation as idempotent or not, when really it's at a field level.  E.g.:

```
doc.update({lastUpdated: FieldValue.serverTimestamp(), count: FieldValue.numericAdd(1)});
```

We only need to store a base state value for 'count', but as written, we'll end up storing base values for 'lastUpdated' and 'count'.  

This is probably non-harmful, but if we want to "fix" this, I think we'd have a getNonIdempotentFields() method instead of getFieldMask() and isIdempotent().

I don't think I care strongly though.</s>

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java [hunk 2]
```diff
@@ -182,16 +187,46 @@ private void startMutationQueue() {
...
+        () -> {
+
+          // For non-idempotent mutations (such as `FieldValue.numericAdd()`), we record the base
+          // state in a separate patch mutation. This is later used to guarantee consistent values
+          // and prevents flicker even if the backend sends us an update that already includes our
+          // transform.
+          List<Mutation> baseMutations = new ArrayList<>();
+          for (Mutation mutation : mutations) {
+            MaybeDocument maybeDocument = existingDocuments.get(mutation.getKey());
+            if (maybeDocument instanceof Document && !mutation.isIdempotent()) {
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java]: What is the latency compensation behavior of:
```
db.collection('votes/michael').set({count: FieldValue.numericAdd(1)}, {merge: true})
```
given votes/michael is not already in cache?

I think I'd expect to get a latency compensated event with {count: 1} in it (do we?).  Assuming we do, I think this code as written won't write a baseValue for 'count' (since we don't have one).  And so it'll be subject to flicker if you receive the watch update before the write ack.

That said, in order for you not to have votes/michael in cache, you can't have already listened to it. So perhaps the only way this could come up is if you start listening while offline (so we'll get latency compensation but it won't be in cache yet), do the merge-set increment, and then come online, and get the watch update before the write ack.  Maybe this is enough of an edge case we don't care that much.

Thoughts?  I can't think of an easy way to fix this unfortunately.  If we care, we may need to re-think the baseMutations thing.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/MutationQueue.java [hunk 3]
```diff
@@ -47,8 +47,17 @@
...
 
-  /** Creates a new mutation batch and adds it to this mutation queue. */
-  MutationBatch addMutationBatch(Timestamp localWriteTime, List<Mutation> mutations);
+  /**
+   * Creates a new mutation batch and adds it to this mutation queue.
+   *
+   * @param localWriteTime The original write time of this mutation.
+   * @param baseMutations Mutations that are used to populate the base values when this mutation is
+   *     applied locally. These mutations are used to locally overwrite values that are persisted in
+   *     the remote document cache.
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/MutationQueue.java]: Consider revising this comment to (at least somewhat) match the proto changes, etc. too?  (and rename baseMutations => baseStateMutations)

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java [hunk 4]
```diff
@@ -56,6 +59,13 @@
 /** A set of helper methods for tests */
 public class IntegrationTestUtil {
 
+  // Whether the integration tests should run against a local Firestore emulator instead of the
+  // Production environment. Note that the Android Emulator treats "10.0.2.2" as its host machine.
+  private static final boolean CONNECT_TO_EMULATOR = false;
+
+  private static final String EMULATOR_HOST = "10.0.2.2";
+  private static final int EMULATOR_PORT = 8081;
```
Comment (mikelehen) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java]: FWIW, I think @allisonbm92 intended us to have multiple FirestoreProvider implementations representing different test environments.  That's how it was before the migration to github.  I don't have strong feelings about this, but we may want to clean this up at some point and either follow the FirestoreProvider pattern, or just ditch FirestoreProvider entirely.

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java [hunk 5]
```diff
@@ -93,11 +111,33 @@ public static FirebaseFirestoreSettings newTestSettings() {
...
-        .setHost(provider.firestoreHost())
-        .setPersistenceEnabled(true)
-        .setTimestampsInSnapshotsEnabled(enabled)
-        .build();
+    FirebaseFirestoreSettings.Builder settings = new FirebaseFirestoreSettings.Builder();
+
+    if (CONNECT_TO_EMULATOR) {
+      settings.setHost(String.format("%s:%d", EMULATOR_HOST, EMULATOR_PORT));
+
+      // Disable SSL and hostname verification
```
Comment (mikelehen) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java]: Fascinating!

This seems strange though. Don't we have a setting to disable SSL, specifically for this sort of test environment?  It may not actually work, but it seems like we should either make it work (by doing something like this?) or else remove that setting and make this the "correct" way to test against a non-ssl environment.

cc/ @allisonbm92 again.

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/NumericTransformsTest.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+import com.google.firebase.firestore.testutil.EventAccumulator;
+import com.google.firebase.firestore.testutil.IntegrationTestUtil;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+
+@Ignore("Not yet available in production")
```
Comment (mikelehen) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/NumericTransformsTest.java]: Can you add a TODO (pointing to you or a bug #) to enable these?

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/NumericTransformsTest.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    assertEquals(expectedSum, (long) snap.getLong("sum"));
+    snap = accumulator.awaitRemoteEvent();
+    assertEquals(expectedSum, (long) snap.getLong("sum"));
+  }
+
+  @Test
+  public void createDocumentWithIncrement() {
+    waitFor(docRef.set(map("sum", FieldValue.numericAdd(1337))));
+    expectLocalAndRemoteValue(1337L);
+  }
```
Comment (mikelehen) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/NumericTransformsTest.java]: Can you do a merge set test too?

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/NumericTransformsTest.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+  @Test
+  public void integerIncrementExistingInteger() {
+    writeInitialData(map("sum", 1337L));
+    waitFor(docRef.update("sum", FieldValue.numericAdd(1)));
+    expectLocalAndRemoteValue(1338L);
+  }
+
+  @Test
+  public void doubleIncrementWithExistingDouble() {
```
Comment (mikelehen) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/NumericTransformsTest.java]: s/With// for consistency with other test names?

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java [hunk 9]
```diff
@@ -934,4 +949,143 @@ public void testRemoteDocumentKeysForTarget() {
...
+    assertChanged(doc("foo/bar", 0, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
+
+    writeMutation(transformMutation("foo/bar", map("sum", FieldValue.numericAdd(2))));
+    assertContains(doc("foo/bar", 0, map("sum", 3), Document.DocumentState.LOCAL_MUTATIONS));
+    assertChanged(doc("foo/bar", 0, map("sum", 3), Document.DocumentState.LOCAL_MUTATIONS));
+  }
+
+  @Test
+  public void testHandlesSetMutationThenAckThenTransformMutationThenAckThenTransformMutation() {
+    if (garbageCollectorIsEager()) {
```
Comment (mikelehen) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: Can you comment why you need this.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java [hunk 10]
```diff
@@ -934,4 +949,143 @@ public void testRemoteDocumentKeysForTarget() {
...
+    assertContains(doc("foo/bar", 0, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
+    assertChanged(doc("foo/bar", 0, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
+
+    writeMutation(transformMutation("foo/bar", map("sum", FieldValue.numericAdd(2))));
+    assertContains(doc("foo/bar", 0, map("sum", 3), Document.DocumentState.LOCAL_MUTATIONS));
+    assertChanged(doc("foo/bar", 0, map("sum", 3), Document.DocumentState.LOCAL_MUTATIONS));
+  }
+
+  @Test
+  public void testHandlesSetMutationThenAckThenTransformMutationThenAckThenTransformMutation() {
```
Comment (mikelehen) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: Did you consider what it would take to be able to make these tests spec tests?  I think that'd be preferable but I think spec tests don't currently do transforms? :-/

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java [hunk 11]
```diff
@@ -934,4 +949,143 @@ public void testRemoteDocumentKeysForTarget() {
...
     assertSetEquals(asList(key("foo/bar"), key("foo/baz")), keys);
   }
+
+  @Test
+  public void testHandlesSetMutationThenTransformMutationThenTransformMutation() {
+    writeMutation(setMutation("foo/bar", map("sum", 0)));
+    assertContains(doc("foo/bar", 0, map("sum", 0), Document.DocumentState.LOCAL_MUTATIONS));
+    assertChanged(doc("foo/bar", 0, map("sum", 0), Document.DocumentState.LOCAL_MUTATIONS));
+
+    writeMutation(transformMutation("foo/bar", map("sum", FieldValue.numericAdd(1))));
```
Comment (mikelehen) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: FWIW, the actual client will never write a batch with just a transform mutation. It'll always be preceded by a patch or set mutation...  So this test is slightly non-realistic. I don't care strongly though.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java (hunk 2 above)
Comment (mikelehen) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java]: LOL 😛 

But now what if instead of a merge-set you do an update() when you have no cached document?

  db.collection('votes/michael').update({count: FieldValue.numericAdd(1)})

I think this will now show up in latency compensation, when it should not (an update() without a base document doesn't get applied to the local view since we don't want to present you with a partial document unless you've specifically opted into this by doing a merge-set instead).

So I'm not sure your fix is right...  I think to get this right we'd only create baseMutations if a) we have a base document, or b) there's a Precondition=None mutation prior to the non-idempotent mutation...  or something like that.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java [hunk 12]
```diff
@@ -123,6 +123,10 @@ public MaybeDocument applyToLocalView(
 
   @Override
   public FieldMask getFieldMask() {
+    // Note: Theoretically, we should only include non-idempotent fields in this field mask as this
+    // mask is used to populate the base state for all DocumentTransforms.  By including  all
+    // fields, we incorrectly prevent rebasing of idempotent transforms (such as `arrayUnion()`)
+    // when any non-idempotent transforms are present.
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java]: This is a good comment, but it may not make sense out-of-context (it assumes knowledge of how this method is going to be used).  I'd actually move it into LocalStore where getFieldMask() is called rather than having it here.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java [hunk 13]
```diff
@@ -968,6 +968,9 @@ public void testHandlesSetMutationThenTransformMutationThenTransformMutation() {
   @Test
   public void testHandlesSetMutationThenAckThenTransformMutationThenAckThenTransformMutation() {
     if (garbageCollectorIsEager()) {
+      // Since this test doesn't open a Query, Eager GC removes the documents from the cache as soon
```
Comment (mikelehen) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: "open a Query" => "start a listen" ?

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java (hunk 10 above)
Comment (mikelehen) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: Maybe add:

// TODO(mrschmidt): These tests would probably be better as spec tests but currently they don't support transforms.

or something?

**Comment by mikelehen** [comment]
@wilhuff Can you review just the local mutation proto change (https://github.com/firebase/firebase-android-sdk/pull/105/files#diff-8ad3f947897b3a40e0037f40aa6c8b4d) and verify you're a fan of this approach (I think you / Sebastian already discussed this some; so just verify it matches your expectations please)?

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java (hunk 5 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/androidTest/java/com/google/firebase/firestore/testutil/IntegrationTestUtil.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java [hunk 14]
```diff
@@ -177,10 +225,19 @@ public boolean isTombstone() {
...
+  /** @return The user-provided mutations in this mutation batch. */
   public List<Mutation> getMutations() {
     return mutations;
   }
+
+  /**
+   * @return The mutations that are used to populate the base values when this mutation batch is
+   *     applied locally.
+   */
+  public List<Mutation> getBaseMutations() {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java]: Hm, I updated the code to use `baseStateMutations`. I left the Proto as is though. I still kind of prefer how `base_writes` is analogous to `writes`.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java (hunk 12 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java]: Yeah, that makes sense. It's basically the same reason as to why I don't want to exclude the idempotent fields here.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java (hunk 1 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java]: We may also not rebase ArrayUnion/ArrayRemove transforms momentarily.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java (hunk 10 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: Done

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java (hunk 13 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java (hunk 2 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java]: I improved my fix further by re-using the same Precondition in the base mutation. I think that should work... Hopefully. Otherwise I have to go back to the drawing board.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java [hunk 15]
```diff
@@ -1088,4 +1094,40 @@ public void testHandlesSetMutationThenTransformMutationThenRemoteEventThenTransf
...
+
+    writeMutations(
+        asList(
+            patchMutation("foo/bar", map()),
+            transformMutation("foo/bar", map("sum", FieldValue.numericAdd(1)))));
+    assertChanged(deletedDoc("foo/bar", 0));
+    assertNotContains("foo/bar");
+
+    applyRemoteEvent(addedRemoteEvent(doc("foo/bar", 1, map("sum", 1337)), asList(2), emptyList()));
+    assertChanged(doc("foo/bar", 1, map("sum", 1), Document.DocumentState.LOCAL_MUTATIONS));
```
Comment (mikelehen) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: FWIW, I don't think this is actually ideal behavior.  Since we won't have raised an initial event with sum=1, it would be better to apply the increment to the actual remote value here and raise sum=1338.

I think ideally we'd somehow implement this so that the _first_ time we apply the Transform and it generates a document (because we had a base document to act on), we'd freeze the base state values and use them from then on until the mutation is acknowledged.  But the current implementation doesn't lend itself to that behavior, so I think we probably want to just live with this as an edge case.

Perhaps add a comment of some kind though so if we ever rework this implementation (e.g. during the overlay work) and look at this test, we'll know that this behavior isn't necessarily ideal and we might want to tweak it.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (mikelehen) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java]: Ehh...  I'm really not sure why "state" makes sense in one place but not the other.  I don't want to compromise by doing a partial rename.  Feel free to rename these back to baseMutations if you don't want to rename base_writes.  I think they should be consistent.

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java]: Renamed back to `baseMutations`.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java [hunk 16]
```diff
@@ -208,21 +207,35 @@ public LocalWriteResult writeLocally(List<Mutation> mutations) {
...
+                  baseStateMutations.add(
+                      new PatchMutation(
+                          mutation.getKey(), baseValues, fieldMask, mutation.getPrecondition()));
+                } else {
+                  baseStateMutations.add(
+                      new PatchMutation(
+                          mutation.getKey(),
+                          ObjectValue.emptyObject(),
+                          fieldMask,
+                          mutation.getPrecondition()));
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java]: That makes sense and makes this a little more obvious. Even I understand it now. Updated.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java (hunk 15 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: I added a comment.

**Comment by schmidt-sebastian** [comment]
I addressed the feedback from the last round. I also made some minor changes (including re-ordering of a unit test to have more specific test go later) in https://github.com/firebase/firebase-android-sdk/pull/105/commits/1f98e0a9b9d158da2dcecc439f34fd8754541d36 . This is mostly from doing the Web Port, and addresses the lack of method overloads in JS.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (mikelehen) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java]: Oh... nice.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java (hunk 15 above)
Comment (mikelehen) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java]: Oh... nice.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (mikelehen) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java]: Oh... nice.

**Comment by schmidt-sebastian** [comment]
FYI: Pushed commit to rename this to "increment" as per the API Council.
